pkg/controller/datavolume/snapshot-clone-controller.go

```diff
@@ -0,0 +1,524 @@
+	// currently don't have CRDs in CI with sourceVolumeMode which is pretty new
+	// converting volume mode is possible but has security implications
+
+	// Sizes validation
+	restoreSize := snapshot.Status.RestoreSize
+	if restoreSize == nil {
+		return true, fmt.Errorf("snapshot has no RestoreSize")
+	}
+	targetRequest, hasTargetRequest := pvcSpec.Resources.Requests[corev1.ResourceStorage]
+	allowExpansion := targetStorageClass.AllowVolumeExpansion != nil && *targetStorageClass.AllowVolumeExpansion
+	if hasTargetRequest {
+		// otherwise will just use restoreSize
+		if restoreSize.Cmp(targetRequest) < 0 && !allowExpansion {
+			r.log.V(3).Info("Can't expand restored PVC because SC does not allow expansion, need to fall back to host assisted")
+			return true, nil
+		}
+	}
+
+	if !isCrossNamespaceClone(datavolume) || *bindingMode == storagev1.VolumeBindingImmediate {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (r *SnapshotCloneReconciler) reconcileRestoreSnapshot(log logr.Logger,
+	datavolume *cdiv1.DataVolume,
+	snapshot *snapshotv1.VolumeSnapshot,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	transferName string,
+	syncRes *dataVolumeCloneSyncResult) (reconcile.Result, error) {
+
+	pvcName := datavolume.Name
+
+	if isCrossNamespaceClone(datavolume) {
+		pvcName = transferName
+		result, err := r.doCrossNamespaceClone(log, syncRes, pvcName, datavolume.Spec.Source.Snapshot.Namespace, false, SmartClone)
+		if result != nil {
+			return *result, err
+		}
+	}
+
+	if datavolume.Status.Phase == cdiv1.NamespaceTransferInProgress {
+		return reconcile.Result{}, nil
+	}
+
+	newPvc, err := r.makePvcFromSnapshot(pvcName, datavolume, snapshot, pvcSpec)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	currentRestoreFromSnapshotPvc := &corev1.PersistentVolumeClaim{}
+	if err := r.client.Get(context.TODO(), client.ObjectKeyFromObject(newPvc), currentRestoreFromSnapshotPvc); err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, err
+		}
+		if err := r.client.Create(context.TODO(), newPvc); err != nil {
+			if cc.ErrQuotaExceeded(err) {
+				r.syncDataVolumeStatusPhaseWithEvent(syncRes, cdiv1.Pending, nil,
+					Event{
+						eventType: corev1.EventTypeWarning,
+						reason:    cc.ErrExceededQuota,
+						message:   err.Error(),
+					})
+			}
+			return reconcile.Result{}, err
+		}
+	} else {
+		if currentRestoreFromSnapshotPvc.Status.Phase == corev1.ClaimBound {
+			if err := r.setCloneOfOnPvc(currentRestoreFromSnapshotPvc); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	return reconcile.Result{}, r.syncCloneStatusPhase(syncRes, cdiv1.CloneFromSnapshotSourceInProgress, nil)
+}
+
+func (r *SnapshotCloneReconciler) createTempHostAssistedSourcePvc(dv *cdiv1.DataVolume, snapshot *snapshotv1.VolumeSnapshot, targetPvcSpec *corev1.PersistentVolumeClaimSpec, syncRes *dataVolumeCloneSyncResult) error {
+	tempPvcName := getTempHostAssistedSourcePvcName(dv.Name)
```

Comment:
`dv.Name` is not unique enough.  What if `ns1/fred` and `ns2/fred` are being cloned concurrently?

Reply:
yup, thx, handled (dv.UID)